lib/segment/src/index/field_index/geo_hash.rs

```diff
@@ -18,6 +31,107 @@ const LON_RANGE: Range<f64> = -180.0..180.0;
 const LAT_RANGE: Range<f64> = -90.0..90.0;
 const COORD_EPS: f64 = 1e-12;
 
+impl Index<usize> for GeoHash {
+    type Output = char;
+
+    fn index(&self, i: usize) -> &Self::Output {
+        assert!(i < self.len());
+        let index = (self.packed >> Self::shift_value(i)) & 0b11111;
+        &BASE32_CODES[index as usize]
+    }
+}
+
+impl From<SmolStr> for GeoHash {
+    fn from(hash: SmolStr) -> Self {
+        Self::new(hash.as_str())
+    }
+}
+
+impl From<GeoHash> for SmolStr {
+    fn from(hash: GeoHash) -> Self {
+        hash.iter().collect()
+    }
+}
+
+pub struct GeoHashIterator {
+    hash: GeoHash,
+    index: usize,
+    len: usize,
+}
+
+impl Iterator for GeoHashIterator {
+    type Item = char;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index < self.len {
+            let result = Some(self.hash[self.index]);
+            self.index += 1;
+            result
+        } else {
+            None
+        }
+    }
+}
+
+impl GeoHash {
+    pub fn new(s: &str) -> Self {
+        assert!(s.len() <= GEOHASH_MAX_LENGTH);
+        let mut packed: u64 = 0;
+        for (i, c) in s.chars().enumerate() {
+            let index = BASE32_CODES.iter().position(|&x| x == c).unwrap() as u64;
+            packed |= index << Self::shift_value(i);
+        }
+        packed |= s.len() as u64;
+        Self { packed }
+    }
+
+    pub fn iter(&self) -> GeoHashIterator {
+        GeoHashIterator {
+            hash: *self,
+            index: 0,
+            len: self.len(),
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn len(&self) -> usize {
+        (self.packed & 0b1111) as usize
+    }
+
+    pub fn cut(&self, new_len: usize) -> Self {
+        assert!(new_len <= self.len());
+        let mut packed = self.packed;
+        for i in new_len..self.len() {
+            packed &= !(0b11111 << Self::shift_value(i));
+        }
```

Comment:
Performance nit: It could be done without a loop:
`packed = !0x1111 << 5 * (self.len() - new_len);` (draft)